### PR TITLE
Eliminate implicit parameters in case classes to avoid implementing `equals()`/ `hashCode()`

### LINKED ISSUE #111: Implement equals() / hashCode() for case classes with implicit parameters
Equals and hashCode are needed to be done manually for case classes with implicit parameters, such as DeserializeContext, DeserializeRegister, ExtractRegisterAs etc

### PR DESCRIPTION
**Pull request by greenhat**
Closes #111

### CODE REVIEW

File: src/main/scala/sigmastate/lang/syntax/Identifiers.scala [hunk 1]
```diff
@@ -6,8 +6,10 @@ import sigmastate.lang.syntax.Basic._
 
 //noinspection ForwardReference
 object Identifiers {
-  case class NamedFunction(f: Char => Boolean)
-      (implicit name: sourcecode.Name) extends (Char => Boolean){
+  case class NamedFunction(f: Char => Boolean,
+                           name: sourcecode.Name = sourcecode.Name.generate)
+    extends (Char => Boolean){
+
```
Comment (greenhat) [inline src/main/scala/sigmastate/lang/syntax/Identifiers.scala]: @ergomorphic I'm not sure the default value is much better here.

**Comment by greenhat** [comment]
@ergomorphic I found only two case classes in our codebase with implicit parameters.

File: src/main/scala/sigmastate/lang/syntax/Identifiers.scala (hunk 1 above)
Comment (ergomorphic) [reply src/main/scala/sigmastate/lang/syntax/Identifiers.scala]: Don’t think we need to change this at all. This issue relate to IR nodes only. Those that are subject to rewriting and thus need structural equality.

File: src/main/scala/sigmastate/Values.scala [hunk 2]
```diff
@@ -424,6 +424,16 @@ object Values {
       case _ =>
         illegalArgs("ConcreteCollection", "(IndexedSeq, SType)", cs)
     }
+
+    override def equals(obj: scala.Any): Boolean = obj match {
+      case cc: ConcreteCollection[_] =>
```
Comment (ergomorphic) [inline src/main/scala/sigmastate/Values.scala]: Please check if we can simple move elementType to regular args. We already have ConcreteCollection.apply with implicit arg.

Comment (greenhat) [reply src/main/scala/sigmastate/Values.scala]: Oops, I missed it. Thanks!

**Comment by greenhat** [comment]
@ergomorphic Fixed.

File: src/main/scala/sigmastate/Values.scala [hunk 3]
```diff
@@ -420,13 +420,17 @@ object Values {
...
       case _ =>
         illegalArgs("ConcreteCollection", "(IndexedSeq, SType)", cs)
     }
   }
   object ConcreteCollection {
-    def apply[V <: SType](items: Value[V]*)(implicit tV: V) = new ConcreteCollection(items.toIndexedSeq)
+    def apply[V <: SType](items: Value[V]*)(implicit tV: V) =
+      new ConcreteCollection(items.toIndexedSeq, tV)
+
+    def apply[V <: SType](items: => Seq[Value[V]])(implicit tV: V) =
```
Comment (ergomorphic) [inline src/main/scala/sigmastate/Values.scala]: Why it is by-name?

File: src/main/scala/sigmastate/Values.scala [hunk 4]
```diff
@@ -420,13 +420,17 @@ object Values {
     def deconstruct = immutable.Seq[Any](elementType) ++ items
 
     def reconstruct(cs: immutable.Seq[Any]) = cs match {
-      case Seq(t: SType, vs@_*) => ConcreteCollection[SType](vs.asInstanceOf[Seq[Value[V]]].toIndexedSeq)(t)
+      case Seq(t: SType, vs@_*) => new ConcreteCollection[SType](vs.asInstanceOf[Seq[Value[V]]].toIndexedSeq, t)
```
Comment (ergomorphic) [inline src/main/scala/sigmastate/Values.scala]: What about removing Rewritable, it is not necessary any more

File: src/main/scala/sigmastate/Values.scala (hunk 3 above)
Comment (greenhat) [reply src/main/scala/sigmastate/Values.scala]: To trick the compiler and not have "same type after erasure" error with first `apply`. Using `IndexedSeq` triggers the same error with compiler generated "default" `apply` for case class. Kinda ugly, but did not find anything better.

File: src/main/scala/sigmastate/Values.scala (hunk 4 above)
Comment (greenhat) [reply src/main/scala/sigmastate/Values.scala]: Done.

**Pull request merged**